In [2]:
#Importing Pandas using an alias to avoid longer text


import pandas as pd

# Reading Data

There are several ways to read data by using Pandas which are specific for specific types of data formats.

Some of the most common and widely used functions in pandas are:-

1. read_csv :- It is used to read data present in CSV (Comma Seperated Values)

2. read_html :- This are used to read data present in form of tables in webpages

3. read_excel :- This is used to read the data present in excel sheets

In [3]:
#Let us read the data which is in csv format now.

draft1 = pd.read_csv('pokemon_data.csv')

draft1

#                   Name   Type 1  Type 2  HP  Attack  Defense  \
0      1              Bulbasaur    Grass  Poison  45      49       49   
1      2                Ivysaur    Grass  Poison  60      62       63   
2      3               Venusaur    Grass  Poison  80      82       83   
3      3  VenusaurMega Venusaur    Grass  Poison  80     100      123   
4      4             Charmander     Fire     NaN  39      52       43   
..   ...                    ...      ...     ...  ..     ...      ...   
795  719                Diancie     Rock   Fairy  50     100      150   
796  719    DiancieMega Diancie     Rock   Fairy  50     160      110   
797  720    HoopaHoopa Confined  Psychic   Ghost  80     110       60   
798  720     HoopaHoopa Unbound  Psychic    Dark  80     160       60   
799  721              Volcanion     Fire   Water  80     110      120   

     Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0         65       65     45           1      False  
1         80       80     60           1      False  
2        100      100     80           1      False  
3        122      120     80           1      False  
4         60       50     65           1      False  
..       ...      ...    ...         ...        ...  
795      100      150     50           6       True  
796      160      110    110           6       True  
797      150      130     70           6       True  
798      170      130     80           6       True  
799      130       90     70           6       True  

[800 rows x 12 columns]

There are two ways in which we can read the data in Pandas by using the read function.

Method 1. If the notebook we are working on is in the same directory as that of the dataset, we can directly read the dataset without    specifying the path

Method 2. If the notebook is present in a different directory, we can use the entire file path to read the dataset. Let us see           how to    do this below. Note that the default path has '/' as the seperators so we can use raw string to read or we           can change '/' to    '\' 

In [4]:
#This is the example for the method 2 mentioned above

example = pd.read_csv(r'C:\Users\Rahul M\Documents\Pandas\survey_results_public.csv')

example

Respondent                                         MainBranch Hobbyist  \
0               1             I am a student who is learning to code      Yes   
1               2             I am a student who is learning to code       No   
2               3  I am not primarily a developer, but I write co...      Yes   
3               4                     I am a developer by profession       No   
4               5                     I am a developer by profession      Yes   
...           ...                                                ...      ...   
88878       88377                                                NaN      Yes   
88879       88601                                                NaN       No   
88880       88802                                                NaN       No   
88881       88816                                                NaN       No   
88882       88863                                                NaN      Yes   

                                             OpenSourcer  \
0                                                  Never   
1                                Less than once per year   
2                                                  Never   
3                                                  Never   
4                             Once a month or more often   
...                                                  ...   
88878  Less than once a month but more than once per ...   
88879                                              Never   
88880                                              Never   
88881                                              Never   
88882                            Less than once per year   

                                              OpenSource  \
0      The quality of OSS and closed source software ...   
1      The quality of OSS and closed source software ...   
2      The quality of OSS and closed source software ...   
3      The quality of OSS and closed source software ...   
4      OSS is, on average, of HIGHER quality than pro...   
...                                                  ...   
88878  The quality of OSS and closed source software ...   
88879  The quality of OSS and closed source software ...   
88880                                                NaN   
88881  OSS is, on average, of HIGHER quality than pro...   
88882  OSS is, on average, of HIGHER quality than pro...   

                                              Employment  \
0                 Not employed, and not looking for work   
1                     Not employed, but looking for work   
2                                     Employed full-time   
3                                     Employed full-time   
4                                     Employed full-time   
...                                                  ...   
88878             Not employed, and not looking for work   
88879                                                NaN   
88880                                 Employed full-time   
88881  Independent contractor, freelancer, or self-em...   
88882             Not employed, and not looking for work   

                      Country         Student  \
0              United Kingdom              No   
1      Bosnia and Herzegovina  Yes, full-time   
2                    Thailand              No   
3               United States              No   
4                     Ukraine              No   
...                       ...             ...   
88878                  Canada              No   
88879                     NaN             NaN   
88880                     NaN             NaN   
88881                     NaN             NaN   
88882                   Spain  Yes, full-time   

                                                 EdLevel  \
0                              Primary/elementary school   
1      Secondary school (e.g. American high school, G...   
2               Bachelor’s degree (BA, BS, B.Eng., etc.)   
3               Bachelor’s degree (BA, BS,

# Head and Tail Function

Now that we have succesfully read the data, the first step would be to take a look at the data to understand the contents of it and get an idea of the basic information present in the data like Name of Columns, No. of Rows etc


This can be achieved by using the Head and Tail functions in Pandas

In [5]:
draft1.head()

#                   Name Type 1  Type 2  HP  Attack  Defense  Sp. Atk  \
0  1              Bulbasaur  Grass  Poison  45      49       49       65   
1  2                Ivysaur  Grass  Poison  60      62       63       80   
2  3               Venusaur  Grass  Poison  80      82       83      100   
3  3  VenusaurMega Venusaur  Grass  Poison  80     100      123      122   
4  4             Charmander   Fire     NaN  39      52       43       60   

   Sp. Def  Speed  Generation  Legendary  
0       65     45           1      False  
1       80     60           1      False  
2      100     80           1      False  
3      120     80           1      False  
4       50     65           1      False

As we can see, the head() function returns the top 5 rows present in the dataset. This is a useful function because we can quickly get an idea of the contents inside the dataset

Similarly, tail() function returns the last 5 rows by default

In [7]:
draft1.tail(10)

#                 Name   Type 1  Type 2   HP  Attack  Defense  Sp. Atk  \
790  714               Noibat   Flying  Dragon   40      30       35       45   
791  715              Noivern   Flying  Dragon   85      70       80       97   
792  716              Xerneas    Fairy     NaN  126     131       95      131   
793  717              Yveltal     Dark  Flying  126     131       95      131   
794  718     Zygarde50% Forme   Dragon  Ground  108     100      121       81   
795  719              Diancie     Rock   Fairy   50     100      150      100   
796  719  DiancieMega Diancie     Rock   Fairy   50     160      110      160   
797  720  HoopaHoopa Confined  Psychic   Ghost   80     110       60      150   
798  720   HoopaHoopa Unbound  Psychic    Dark   80     160       60      170   
799  721            Volcanion     Fire   Water   80     110      120      130   

     Sp. Def  Speed  Generation  Legendary  
790       40     55           6      False  
791       80    123           6      False  
792       98     99           6       True  
793       98     99           6       True  
794       95     95           6       True  
795      150     50           6       True  
796      110    110           6       True  
797      130     70           6       True  
798      130     80           6       True  
799       90     70           6       True

If the value is not specified, head() and tail() returns the first and the last 5 rows respecitvely but if the values are specified, they return the suitable number of rows accordingly as shown above

There are many parameters which are available inside read_csv() so lets look at a few of them right now

1. header = 'infer'
2. names = None
3. index_col = None



## header = 'infer' and names = None

By default, when we read any dataset, the parameter header is set to be 'infer'. Infer stands for inferring the header names from the dataset itself.

If no names are passed as column names in the read_csv() function, header becomes equal to 'infer' and Pandas automatically pass in the first row elements as column names by default. In this case, names are set to None

If names are passed explicitly, header becomes equal to none and those names are used as Column Names.

Let us see this with the help of examples

In [14]:
draft2 = pd.read_csv('pokemon_data.csv')

Here, the names of the columns are not explicitly mentioned, so header is equal to infer and column names are the first row values of the dataset

In [15]:
colnames = list(draft1.columns)
colnames.append('New Name')


draft3 = pd.read_csv('pokemon_data.csv',names=colnames)


In [16]:
draft3

#                   Name   Type 1  Type 2  HP  Attack  Defense  \
0      #                   Name   Type 1  Type 2  HP  Attack  Defense   
1      1              Bulbasaur    Grass  Poison  45      49       49   
2      2                Ivysaur    Grass  Poison  60      62       63   
3      3               Venusaur    Grass  Poison  80      82       83   
4      3  VenusaurMega Venusaur    Grass  Poison  80     100      123   
..   ...                    ...      ...     ...  ..     ...      ...   
796  719                Diancie     Rock   Fairy  50     100      150   
797  719    DiancieMega Diancie     Rock   Fairy  50     160      110   
798  720    HoopaHoopa Confined  Psychic   Ghost  80     110       60   
799  720     HoopaHoopa Unbound  Psychic    Dark  80     160       60   
800  721              Volcanion     Fire   Water  80     110      120   

     Sp. Atk  Sp. Def  Speed  Generation  Legendary  New Name  
0    Sp. Atk  Sp. Def  Speed  Generation  Legendary       NaN  
1         65       65     45           1      FALSE       NaN  
2         80       80     60           1      FALSE       NaN  
3        100      100     80           1      FALSE       NaN  
4        122      120     80           1      FALSE       NaN  
..       ...      ...    ...         ...        ...       ...  
796      100      150     50           6       TRUE       NaN  
797      160      110    110           6       TRUE       NaN  
798      150      130     70           6       TRUE       NaN  
799      170      130     80           6       TRUE       NaN  
800      130       90     70           6       TRUE       NaN  

[801 rows x 13 columns]

Here, the column names are explicitly mentioned and passed in the read_csv() function so this time header becomes equal to None and these names are used as column names. 

Notice that the usual column names occupied the first row this time

## index_col

As the name implies, index_col refers to the index column of the dataframe. By default, index_col is set to None and indexing usually starts from 0 in this case till the end of number of rows.

There are two ways in which we can index the dataframe in Pandas

1. Single Indexing
2. Multi Indexing

In [19]:
draft1.head()

#                   Name Type 1  Type 2  HP  Attack  Defense  Sp. Atk  \
0  1              Bulbasaur  Grass  Poison  45      49       49       65   
1  2                Ivysaur  Grass  Poison  60      62       63       80   
2  3               Venusaur  Grass  Poison  80      82       83      100   
3  3  VenusaurMega Venusaur  Grass  Poison  80     100      123      122   
4  4             Charmander   Fire     NaN  39      52       43       60   

   Sp. Def  Speed  Generation  Legendary  
0       65     45           1      False  
1       80     60           1      False  
2      100     80           1      False  
3      120     80           1      False  
4       50     65           1      False

In [20]:
ex0 = pd.read_csv('pokemon_data.csv',index_col='#')

ex0.head()

Name Type 1  Type 2  HP  Attack  Defense  Sp. Atk  \
#                                                                       
1              Bulbasaur  Grass  Poison  45      49       49       65   
2                Ivysaur  Grass  Poison  60      62       63       80   
3               Venusaur  Grass  Poison  80      82       83      100   
3  VenusaurMega Venusaur  Grass  Poison  80     100      123      122   
4             Charmander   Fire     NaN  39      52       43       60   

   Sp. Def  Speed  Generation  Legendary  
#                                         
1       65     45           1      False  
2       80     60           1      False  
3      100     80           1      False  
3      120     80           1      False  
4       50     65           1      False

Notice the differences in draft1 and ex0 dataframes.

In draft1, we did not specify the index column so the indexing started from 0 and continues onwards till the end.

But in ex0, since we explicitly mentioned index_col = '#' column, the '#' column is set as index so in this case, indexing started from 1.

This is an example of Single Indexing

In [22]:
ex1 = pd.read_csv('pokemon_data.csv',index_col = ['#','Legendary'])

ex1.head(20)

Name Type 1  Type 2  HP  Attack  Defense  \
#  Legendary                                                                  
1  False                      Bulbasaur  Grass  Poison  45      49       49   
2  False                        Ivysaur  Grass  Poison  60      62       63   
3  False                       Venusaur  Grass  Poison  80      82       83   
   False          VenusaurMega Venusaur  Grass  Poison  80     100      123   
4  False                     Charmander   Fire     NaN  39      52       43   
5  False                     Charmeleon   Fire     NaN  58      64       58   
6  False                      Charizard   Fire  Flying  78      84       78   
   False      CharizardMega Charizard X   Fire  Dragon  78     130      111   
   False      CharizardMega Charizard Y   Fire  Flying  78     104       78   
7  False                       Squirtle  Water     NaN  44      48       65   
8  False                      Wartortle  Water     NaN  59      63       80   
9  False                      Blastoise  Water     NaN  79      83      100   
   False        BlastoiseMega Blastoise  Water     NaN  79     103      120   
10 False                       Caterpie    Bug     NaN  45      30       35   
11 False                        Metapod    Bug     NaN  50      20       55   
12 False                     Butterfree    Bug  Flying  60      45       50   
13 False                         Weedle    Bug  Poison  40      35       30   
14 False                         Kakuna    Bug  Poison  45      25       50   
15 False                       Beedrill    Bug  Poison  65      90       40   
   False          BeedrillMega Beedrill    Bug  Poison  65     150       40   

              Sp. Atk  Sp. Def  Speed  Generation  
#  Legendary                                       
1  False           65       65     45           1  
2  False           80       80     60           1  
3  False          100      100     80           1  
   False          122      120     80           1  
4  False           60       50     65           1  
5  False           80       65     80           1  
6  False          109       85    100           1  
   False          130       85    100           1  
   False          159      115    100           1  
7  False           50       64     43           1  
8  False           65       80     58           1  
9  False           85      105     78           1  
   False          135      115     78           1  
10 False           20       20     45           1  
11 False           25       25     30           1  
12 False           90       80     70           1  
13 False           20       20     50           1  
14 False           25       25     35           1  
15 False           45       80     75           1  
   False           15       80    145           1

ex2 dataframe is an example of Multi Indexing wherein the dataframe has two indexes namely '#' and 'Legendary'

These are the various types of indexing we can use in Pandas

# isnull() function

The isnull() function is one of the most used functions while working with Pandas Dataframes. Often the data we obtain from the real-world is messy and it has a lot of NaN or Null Values that comes along with it. 

The isnull() function returns the Dataframe consisting of Trues and Falses based on if the data present in that location is null or not.

Let us see the example

In [23]:
draft1.isnull()

#   Name  Type 1  Type 2     HP  Attack  Defense  Sp. Atk  Sp. Def  \
0    False  False   False   False  False   False    False    False    False   
1    False  False   False   False  False   False    False    False    False   
2    False  False   False   False  False   False    False    False    False   
3    False  False   False   False  False   False    False    False    False   
4    False  False   False    True  False   False    False    False    False   
..     ...    ...     ...     ...    ...     ...      ...      ...      ...   
795  False  False   False   False  False   False    False    False    False   
796  False  False   False   False  False   False    False    False    False   
797  False  False   False   False  False   False    False    False    False   
798  False  False   False   False  False   False    False    False    False   
799  False  False   False   False  False   False    False    False    False   

     Speed  Generation  Legendary  
0    False       False      False  
1    False       False      False  
2    False       False      False  
3    False       False      False  
4    False       False      False  
..     ...         ...        ...  
795  False       False      False  
796  False       False      False  
797  False       False      False  
798  False       False      False  
799  False       False      False  

[800 rows x 12 columns]

Just by looking at this dataframe, we can see a clear shortcoming. Although this does return the true or falses, since the data is huge, we cannot clearly see how many columns have null values

To overcome this problem, we use sum() function which returns the number of null values in each column of the dataframe

In [24]:
draft1.isnull().sum()

#               0
Name            0
Type 1          0
Type 2        386
HP              0
Attack          0
Defense         0
Sp. Atk         0
Sp. Def         0
Speed           0
Generation      0
Legendary       0
dtype: int64

As we can see, Type 2 Column in this dataframe has 386 null values.

## Excel Reading Syntax

In order to read the data from excel sheets, we follow a usual syntax which looks like this.

pd.read_excel('name.xlsx',sheet_name='name')

Since one excel file can have multiple sheets, we use the sheet name to read the data from specific sheet

In [26]:
excel1 = pd.read_excel(r'C:\Users\Rahul M\Documents\Book1.xlsx',sheet_name='Details')

excel1

Name  Age  Salary
0  James   23   12000
1  Jason   21   23000
2    Tim   24   30000

## Reading from URL

Let us read the data from a url which contains the data in the form of tables and not as raw data. Inorder to read the data from such sources, we use read_html() function to return the table of data


In [27]:
url = 'https://www.basketball-reference.com/leagues/NBA_2015_totals.html'

exa1 = pd.read_html(url)

exa1

[      Rk          Player Pos Age   Tm   G  GS    MP   FG  FGA  ...   FT%  ORB  \
 0      1      Quincy Acy  PF  24  NYK  68  22  1287  152  331  ...  .784   79   
 1      2    Jordan Adams  SG  20  MEM  30   0   248   35   86  ...  .609    9   
 2      3    Steven Adams   C  21  OKC  70  67  1771  217  399  ...  .502  199   
 3      4     Jeff Adrien  PF  28  MIN  17   0   215   19   44  ...  .579   23   
 4      5   Arron Afflalo  SG  29  TOT  78  72  2502  375  884  ...  .843   27   
 ..   ...             ...  ..  ..  ...  ..  ..   ...  ...  ...  ...   ...  ...   
 670  490  Thaddeus Young  PF  26  TOT  76  68  2434  451  968  ...  .655  127   
 671  490  Thaddeus Young  PF  26  MIN  48  48  1605  289  641  ...  .682   75   
 672  490  Thaddeus Young  PF  26  BRK  28  20   829  162  327  ...  .606   52   
 673  491     Cody Zeller   C  22  CHO  62  45  1487  172  373  ...  .774   97   
 674  492    Tyler Zeller   C  25  BOS  82  59  1731  340  619  ...  .823  146   
 
      DRB  TRB

In [29]:
type(exa1)

list

Even though the code worked properly, we can see that the data is not present in the form of a dataframe, instead it is in the form of a list. 

This is because of the fact that, when we use read_html() function on a url, it returns the data which are present in tables. These tables can be one or many depending on the website.

But, since it is difficult to display all the tables at once, this function appends all the dataframes in a list, essentially wrapping them.

So inorder to access individual dataframes, we need to index the list

In [30]:
dfa = exa1[0]
dfa

Rk          Player Pos Age   Tm   G  GS    MP   FG  FGA  ...   FT%  ORB  \
0      1      Quincy Acy  PF  24  NYK  68  22  1287  152  331  ...  .784   79   
1      2    Jordan Adams  SG  20  MEM  30   0   248   35   86  ...  .609    9   
2      3    Steven Adams   C  21  OKC  70  67  1771  217  399  ...  .502  199   
3      4     Jeff Adrien  PF  28  MIN  17   0   215   19   44  ...  .579   23   
4      5   Arron Afflalo  SG  29  TOT  78  72  2502  375  884  ...  .843   27   
..   ...             ...  ..  ..  ...  ..  ..   ...  ...  ...  ...   ...  ...   
670  490  Thaddeus Young  PF  26  TOT  76  68  2434  451  968  ...  .655  127   
671  490  Thaddeus Young  PF  26  MIN  48  48  1605  289  641  ...  .682   75   
672  490  Thaddeus Young  PF  26  BRK  28  20   829  162  327  ...  .606   52   
673  491     Cody Zeller   C  22  CHO  62  45  1487  172  373  ...  .774   97   
674  492    Tyler Zeller   C  25  BOS  82  59  1731  340  619  ...  .823  146   

     DRB  TRB  AST  STL BLK  TOV   PF   PTS  
0    222  301   68   27  22   60  147   398  
1     19   28   16   16   7   14   24    94  
2    324  523   66   38  86   99  222   537  
3     54   77   15    4   9    9   30    60  
4    220  247  129   41   7  116  167  1035  
..   ...  ...  ...  ...  ..  ...  ...   ...  
670  284  411  173  124  25  117  171  1071  
671  170  245  135   86  17   75  115   685  
672  114  166   38   38   8   42   56   386  
673  265  362  100   34  49   62  156   472  
674  319  465  113   18  52   76  205   833  

[675 rows x 30 columns]

Now we can see that this data is a dataframe

# Selecting one or multiple columns

In Pandas, we can select one or multiple columns by passing their names in a list within the dataframe.

Lets look at how we do it

In [31]:
draft1['Name']

0                  Bulbasaur
1                    Ivysaur
2                   Venusaur
3      VenusaurMega Venusaur
4                 Charmander
               ...          
795                  Diancie
796      DiancieMega Diancie
797      HoopaHoopa Confined
798       HoopaHoopa Unbound
799                Volcanion
Name: Name, Length: 800, dtype: object

In [32]:
type(draft1['Name'])

pandas.core.series.Series

In [33]:
draft1[['Name']]

Name
0                Bulbasaur
1                  Ivysaur
2                 Venusaur
3    VenusaurMega Venusaur
4               Charmander
..                     ...
795                Diancie
796    DiancieMega Diancie
797    HoopaHoopa Confined
798     HoopaHoopa Unbound
799              Volcanion

[800 rows x 1 columns]

In [34]:
type(draft1[['Name']])

pandas.core.frame.DataFrame

As we can notice from the past 4 cells, if we mention the name of the column within a single [], pandas interpret this data as Pandas Series and not Pandas Dataframe. Inorder for us to make this into a dataframe, we have to pass it as a list with the [].

The same is the case for multiple columns selecting. If we pass multiple column names without a list, we will encounter an error since a Pandas Series can only have one column

In [35]:
draft1[['Name','Type 1']]

Name   Type 1
0                Bulbasaur    Grass
1                  Ivysaur    Grass
2                 Venusaur    Grass
3    VenusaurMega Venusaur    Grass
4               Charmander     Fire
..                     ...      ...
795                Diancie     Rock
796    DiancieMega Diancie     Rock
797    HoopaHoopa Confined  Psychic
798     HoopaHoopa Unbound  Psychic
799              Volcanion     Fire

[800 rows x 2 columns]

## .columns

Inorder to quickly look at the number of columns in a dataframe, we can use the .columns function.

This function returns all the column names of the dataframe in a list format

In [36]:
draft1.columns

Index(['#', 'Name', 'Type 1', 'Type 2', 'HP', 'Attack', 'Defense', 'Sp. Atk',
       'Sp. Def', 'Speed', 'Generation', 'Legendary'],
      dtype='object')

## .dtypes

We can navigate through the datatypes of all the columns by using the .dtypes function in Pandas.

In [37]:
draft1.dtypes

#              int64
Name          object
Type 1        object
Type 2        object
HP             int64
Attack         int64
Defense        int64
Sp. Atk        int64
Sp. Def        int64
Speed          int64
Generation     int64
Legendary       bool
dtype: object